File: backend/controller/pubsub/testdata/go/subscriber/subscriber.go
package subscriber

import (
	"context"
	"fmt"
	"ftl/builtin"
	"ftl/publisher"
	"strings"
	"time"

	"github.com/TBD54566975/ftl/go-runtime/ftl"
	"github.com/TBD54566975/ftl/go-runtime/ftl/reflection"
	"github.com/alecthomas/atomic"
	//"github.com/TBD54566975/ftl/go-runtime/ftl" // Import the FTL SDK.
)

type TestTopicSubscription = ftl.SubscriptionHandle[publisher.TestTopic, ConsumeClient, publisher.PubSubEvent]
type DoomedSubscription = ftl.SubscriptionHandle[publisher.Topic2, ConsumeButFailAndRetryClient, publisher.PubSubEvent]
type DoomedSubscription2 = ftl.SubscriptionHandle[publisher.Topic2, ConsumeButFailAndCatchAnyClient, publisher.PubSubEvent]

var catchCount atomic.Value[int]

//ftl:verb
func Consume(ctx context.Context, req publisher.PubSubEvent) error {
	ftl.LoggerFromContext(ctx).Infof("Subscriber is consuming %v", req.Time)
	return nil
}

//ftl:verb
//ftl:retry 2 1s 1s catch catch
func ConsumeButFailAndRetry(ctx context.Context, req publisher.PubSubEvent) error {
	return fmt.Errorf("always error: event %v", req.Time)
}

//ftl:verb
//ftl:retry 1 1s 1s catch catchAny
func ConsumeButFailAndCatchAny(ctx context.Context, req publisher.PubSubEvent) error {
	return fmt.Errorf("always error: event %v", req.Time)
}

//ftl:verb
func PublishToExternalModule(ctx context.Context) error {
	// Get around compile-time checks
	externalTopic := ftl.TopicHandle[publisher.PubSubEvent]{Ref: reflection.Ref{Module: "publisher", Name: "testTopic"}.ToSchema()}
	return externalTopic.Publish(ctx, publisher.PubSubEvent{Time: time.Now()})
}

//ftl:verb
func Catch(ctx context.Context, req builtin.CatchRequest[publisher.PubSubEvent]) error {
	if !strings.Contains(req.Error, "always error: event") {
		return fmt.Errorf("unexpected error: %v", req.Error)
	}
	count := catchCount.Load() + 1
	catchCount.Store(count)

	// fail once
	if count == 1 {
		return fmt.Errorf("catching error")
	}

	// succeed after that
	return nil
}

//ftl:verb
func CatchAny(ctx context.Context, req builtin.CatchRequest[any]) error {
	if req.Verb.Module != "subscriber" {
		return fmt.Errorf("unexpected verb module: %v", req.Verb.Module)
	}
	if req.Verb.Name != "consumeButFailAndCatchAny" {
		return fmt.Errorf("unexpected verb name: %v", req.Verb.Name)
	}
	if req.RequestType != "publisher.PubSubEvent" {
		return fmt.Errorf("unexpected request type: %v", req.RequestType)
	}
	requestMap, ok := req.Request.(map[string]any)
	if !ok {
		return fmt.Errorf("expected request to be a map[string]any: %T", req.Request)
	}
	timeValue, ok := requestMap["time"]
	if !ok {
		return fmt.Errorf("expected request to have a time key")
	}
	if _, ok := timeValue.(string); !ok {
		return fmt.Errorf("expected request to have a time value of type string")
	}
	return nil
}

File: backend/controller/pubsub/testdata/java/subscriber/src/main/java/xyz/block/ftl/java/test/subscriber/Subscriber.java
package xyz.block.ftl.java.test.subscriber;

import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import ftl.builtin.CatchRequest;
import ftl.publisher.PubSubEvent;
import ftl.publisher.TestTopicSubscription;
import io.quarkus.logging.Log;
import xyz.block.ftl.Retry;
import xyz.block.ftl.Subscription;
import xyz.block.ftl.Verb;
import xyz.block.ftl.VerbName;

public class Subscriber {

    private static final AtomicInteger catchCount = new AtomicInteger();

    @TestTopicSubscription
    void consume(PubSubEvent event) throws Exception {
        Log.infof("Subscriber is consuming %s", event.getTime());
    }

    @Subscription(topic = "topic2", module = "publisher", name = "doomedSubscription")
    @Retry(count = 2, minBackoff = "1s", maxBackoff = "1s", catchVerb = "catch")
    public void consumeButFailAndRetry(PubSubEvent event) {
        throw new RuntimeException("always error: event " + event.getTime());
    }

    @Subscription(topic = "topic2", module = "publisher", name = "doomedSubscription2")
    @Retry(count = 1, minBackoff = "1s", maxBackoff = "1s", catchVerb = "catchAny")
    public void consumeButFailAndCatchAny(PubSubEvent event) {
        throw new RuntimeException("always error: event " + event.getTime());
    }

    @Verb
    @VerbName("catch")
    public void catchVerb(CatchRequest<PubSubEvent> req) {
        if (!req.getError().contains("always error: event")) {
            throw new RuntimeException("unexpected error: " + req.getError());
        }
        if (catchCount.incrementAndGet() == 1) {
            throw new RuntimeException("catching error");
        }
    }

    @Verb
    public void catchAny(CatchRequest<Object> req) {
        if (!"subscriber".equals(req.getVerb().getModule())) {
            throw new IllegalArgumentException(String.format("unexpected verb module: %s", req.getVerb().getModule()));
        }
        if (!"consumeButFailAndCatchAny".equals(req.getVerb().getName())) {
            throw new IllegalArgumentException(String.format("unexpected verb name: %s", req.getVerb().getName()));
        }
        if (!"publisher.PubSubEvent".equals(req.getRequestType())) {
            throw new IllegalArgumentException(String.format("unexpected request type: %s", req.getRequestType()));
        }
        if (!(req.getRequest()instanceof Map<?,?>)) {
            throw new IllegalArgumentException(
                    String.format("expected request to be a Map: %s", req.getRequest().getClass().getName()));
        }
        var request = (Map<?, ?>) req.getRequest();
        var time = request.get("time");
        if (time == null) {
            throw new IllegalArgumentException("expected request to have a time key");
        }
        if (!(time instanceof String)) {
            throw new IllegalArgumentException("expected request to have a time value of type string");
        }
    }

}
